### DISCUSSION TITLE
Lesson 7: Gas Estimation Error when Ending Lottery

### DISCUSSION THREAD

**Discussion by abdullahpichoo** [smartcontractkit/full-blockchain-solidity-course-py#1669] [Q&A] [answered]
## Problem
I was working with the Lottery contract and everything went successful when I deployed to a Local Chain. The issue started with the testnet. The contract gives a Gas Estimation Error when the ```endLottery()``` function calls the ```requestRandomness``` function.

I know that my contract must have LINK before it can call that randomness function. I do call the ```fund_with_link``` function and I've checked on Etherscan and I send 1 LINK successfully to my contract (in the ERC20 Token section). The fees for a RandomCall transaction is 0.25 LINK and I think funding my contract with 1 LINK would be sufficient to make the Random Number call but guess what, Gas Estimation failed.

I would really appreciate any help! Thanks!

## Error
<img width="883" alt="image" src="https://user-images.githubusercontent.com/98353306/186782903-ded53e2e-7583-46d0-bf91-029ac0f73675.png">

## 1 LINK Transfer
<img width="678" alt="image" src="https://user-images.githubusercontent.com/98353306/186782971-b808c29b-6fbc-44c4-8ef3-b1149fda8704.png">

## Code
Here is the ```endLottery``` ```fund_with_link``` scripts that are causing the issue.
``` Python

def end_lottery():
    acc = get_acc()
    lottery = Lottery[-1]
    print(f"Balance of this contract before funding: {lottery.balance()}")
    tx = fund_with_link(lottery.address)
    tx.wait(1)
    print(f"Balance of this contract: {lottery.balance()}")
    tx = lottery.endLottery({"from": acc})
    tx.wait(1)
    time.sleep(60)
    winner_address = lottery.last_winner()
    print(
        f"Winner Address is: {winner_address} with Name: {lottery.player_names(winner_address)}"
    )

LINK_AMOUNT = 1000000000000000000


def fund_with_link(contract_address, acc=None, link_token=None, amount=LINK_AMOUNT):
    if acc:
        acc = acc
    else:
        acc = get_acc()

    if link_token:
        link_token = link_token
    else:
        link_token = get_contract("link-token")
    # 7.43.48 Interface working example to replace Line 55
    # link_token_contract = interface.LinkTokenInterface(link_token.address)
    tx = link_token.transfer(contract_address, amount, {"from": acc})
    tx.wait(1)
    print(f"Contract Funded with Link")
    return tx

```

``` Solidity
function endLottery() public {
        state = LOTTERY_STATE.running;
        bytes32 requestID = requestRandomness(keyHash, fee); //This will return a bytes32 requestID
        emit requested_randomness(requestID);
        //This is done so that the request ID is emitted onthe blockchain node as an event that we can access
    }
```

**Comment by cromewar** [comment, marked as answer]
Hello @abdullahpichoo 
Tricky question, but thanks for the detailed explanation of the error this is really helpful.
So I would say there is not taking one of the contracts correctly, be sure the name of `link-token` is setup the same for the network you are deploying to on your `brownie-config.yaml` and that your `get_ contract` function is correct.

Also, I highly recommend to delete the build folder, and re-deploy the contract.. then try again.

Please let us know if this was helpful.

**Reply by abdullahpichoo** [reply]
Fixed this issue. The Problem was in my brownie-config file. As the YT tutorial is using the VRF Coordinator V1, I would need to pass the VRFCoordinatorV1 address to its respective contract. But as the Chainlink VRF website has been updated to V2, they had the address of VRFCoordinatorV2 next to the VRFCoordinator in the Supported Networks tab. So I had to change that out. And yeah I had to change the keyhash as well.  So, in short, I was passing the address of VRFCoordinatorV2 to the contract of VRFCoordinatorV1, which caused this issue.

**Reply by cromewar** [reply]
I'm glad your problem is solved now!

**Reply by Cosmodude** [reply]
Changed the brownie config file to V1 and recompiled it, still have Gas estimation error.

**Reply by abdullahpichoo** [reply]
Hey @hackdady3000 , yes the VRFv1 was causing issues for me too. That's why I implemented the VRFv2 with a custom python script. You can check it here. Hope it helps.
Link: https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1675
